### PR TITLE
Null value Auth#clientId can indicate anonymous or not identified

### PR DESCRIPTION
**Pull request by mattheworiordan**
@paddybyers @SimonWoolf 

Following the closure of https://github.com/ably/docs/pull/71, there are a few small changes needed to the spec.

Please can you review as I hope to distribute the final spec changes to everyone tomorrow.

### CODE REVIEW

**Comment by mattheworiordan** [comment]
FYI, I just added 96f63fd as I think we should clear up terminology now so that we don't have to come back and adjust this spec again.

Feedback appreciated, would like to merge ASAP

File: content/client-lib-development-guide/features.textile [hunk 1]
```diff
@@ -58,7 +58,7 @@ h3(#restclient). RestClient
 * @(RSC16)@ @RestClient#time@ function sends a get request to @rest.ably.io/time@ and returns the server time in milliseconds since epoch or as a Date/Time object where suitable
 * @(RSC7)@ Sends REST requests over HTTP and HTTPS to the REST end-point @rest.ably.io@
 ** @(RSC7a)@ The header @X-Ably-Version: 0.8@ must be included in all REST requests to the Ably end-point
-* @(RSC18)@ If @ClientOptions#tls@ is true, then all communication is over HTTPS. If false, all communication is over HTTP however Basic Auth over HTTP will raise an exception as private keys cannot be submitted over an insecure connection. See @Auth@ below
+* @(RSC18)@ If @ClientOptions#tls@ is true, then all communication is over HTTPS. If false, all communication is over HTTP however "Basic Auth":https://en.wikipedia.org/wiki/Basic_access_authentication over HTTP will raise an exception as private keys cannot be submitted over an insecure connection. See @Auth@ below
```
Comment (paddybyers) [inline content/client-lib-development-guide/features.textile]: > raise an exception

This is environment-specific - I think we were preferring wording such as "result in an error" which is generic.

File: content/client-lib-development-guide/features.textile [hunk 2]
```diff
@@ -81,35 +81,37 @@ h3(#restclient). RestClient
 
 h3(#rest-auth). Auth
 
-* @(RSA1)@ Basic Auth connects over HTTPS by default.  If HTTP and Basic Auth are used, an exception will be raised
+* @(RSA1)@ "Basic Auth":https://en.wikipedia.org/wiki/Basic_access_authentication connects over HTTPS by default.  If HTTP and Basic Auth are used, an exception will be raised
```
Comment (paddybyers) [inline content/client-lib-development-guide/features.textile]: > an exception will be raised

see above

File: content/client-lib-development-guide/features.textile [hunk 3]
```diff
@@ -81,35 +81,37 @@ h3(#restclient). RestClient
...
-*** @(RSA7a1)@ All operations (such as message publishing or presence) will have an implicit @clientId@. The Ably service automatically updates the @clientId@ attribute (when empty) for all @Message@ and @PresenceMessage@ messages received from that authenticated client, and any messages then published from the Ably service, will have the @clientId@ attribute populated. It is therefore expected that Ably client libraries do not explicitly set the @clientId@ field on messages published when @clientId@ is implicit in the connection or authentication scheme
+** @(RSA7a)@ If a @clientId@ is provided in the @ClientOptions@, or is present in the current authentication token, then the client is considered to be identified (it has a @clientId@ that is implicit in all operations). Note that an authentication token @clientId@ wildcard value of @'*'@ is the exception where the client is unidentified (restricted to one identity) and any @clientId@ is permitted. The following applies to identified clients:
```
Comment (paddybyers) [inline content/client-lib-development-guide/features.textile]: > unidentified (restricted to one identity)

I don't understand this comment

File: content/client-lib-development-guide/features.textile [hunk 4]
```diff
@@ -381,15 +383,15 @@ h3(#realtime-channel). Channel
...
+**** @(RTL6g1b)@ Ably will not assign a @clientId@ upon receiving the @Message@. A test should assert that the @clientId@ value is @null@ for the @Message@ when received
+*** @(RTL6g2)@ When publishing a @Message@ with the @clientId@ attribute value set to the identified client's @clientId@, Ably will accept the message and publish it. A test should assert that the @clientId@ value is populated for the @Message@ when received
+*** @(RTL6g3)@ When publishing a @Message@ with a different @clientId@ attribute value to the identified client's @clientId@, the client library should reject publishing immediately. The message should not be sent to Ably and the errr callback should be called. The connection and channel must remain available for further publishing and other operations
```
Comment (paddybyers) [inline content/client-lib-development-guide/features.textile]: > different @clientId@ attribute value to the identified

different .... from ....

> should reject publishing immediately

I think  "... should reject that Message ..." or "... should reject that publish operation ...". The current wording sounds like perhaps no further messages will be published.

File: content/client-lib-development-guide/features.textile [hunk 5]
```diff
@@ -477,7 +479,7 @@ h3(#realtime-presence). Presence
 ** @(RTP15c)@ Tests should also ensure that using these methods has no side effects on a client that has entered normally using @Presence#enter@
 ** @(RTP15d)@ A callback can be provided that will be called upon success or failure
 ** @(RTP15e)@ Implicitly attaches the @Channel@. However, if the channel is in the @FAILED@ state, an exception is raised
-** @(RTP15f)@ If the client is authenticated and has a configured @clientId@, if the @clientId@ argument does not match the client's @clientId@, then it should indicate an error. The connection and channel remains available for further operations
+** @(RTP15f)@ If the client is identified and has a valid @clientId@, and the @clientId@ argument does not match the client's @clientId@, then it should indicate an error. The connection and channel remains available for further operations
```
Comment (paddybyers) [inline content/client-lib-development-guide/features.textile]: > remains

remain

**Comment by paddybyers** [comment]
A few comments, otherwise LGTM.

File: content/client-lib-development-guide/features.textile [hunk 6]
```diff
@@ -81,35 +81,37 @@ h3(#restclient). RestClient
...
-* @(RSA2)@ Basic Auth is the default authentication scheme if an API key exists
+* @(RSA2)@ "Basic Auth":https://en.wikipedia.org/wiki/Basic_access_authentication is the default authentication scheme if an API key exists
 * @(RSA3)@ Token Auth:
 ** @(RSA3a)@ Can be used over HTTP or HTTPs
 ** @(RSA3b)@ For REST requests, the token string is Base64 encoded and used in the @Authorization: Bearer@ header
 ** @(RSA3c)@ For Realtime connections, the querystring param @access_token@ is appended to the URL endpoint
-* @(RSA4)@ Token Auth is used if @useTokenAuth@ is set to true, or if @useTokenAuth@ is unspecified and any one of the following conditions cause token auth to be selected as the default: a @clientId@ is specified; @authUrl@ or @authCallback@ is provided; an explicit @token@ or @TokenDetails@ is provided
+* @(RSA4)@ Token Auth is used if @useTokenAuth@ is set to true, or if @useTokenAuth@ is unspecified and any one of the following conditions cause Token Auth to be selected as the default: a @clientId@ is specified; @authUrl@ or @authCallback@ is provided; an explicit @token@ or @TokenDetails@ is provided
```
Comment (SimonWoolf) [inline content/client-lib-development-guide/features.textile]: Redundant wording: "Token Auth is used if ... cause Token Auth to be selected as the default"
Suggest: `Token Auth is used if @useTokenAuth@ is set to true, or if @useTokenAuth@ is unspecified and any of the following apply: a @clientId@ is specified; @authUrl@ or @authCallback@ is provided; an explicit @token@ or @TokenDetails@ is provided`

File: content/client-lib-development-guide/features.textile (hunk 3 above)
Comment (mattheworiordan) [inline content/client-lib-development-guide/features.textile]: > (restricted to one identity)

It's a typo

File: content/client-lib-development-guide/features.textile [hunk 7]
```diff
@@ -81,35 +81,37 @@ h3(#restclient). RestClient
...
+** @(RSA7a)@ If a @clientId@ is provided in the @ClientOptions@, or is present in the current authentication token, then the client is considered to be identified (it has a @clientId@ that is implicit in all operations). Note that an authentication token @clientId@ wildcard value of @'*'@ is the exception where the client is unidentified (restricted to one identity) and any @clientId@ is permitted. The following applies to identified clients:
+*** @(RSA7a1)@ All operations (such as message publishing or presence) will have an implicit @clientId@. The Ably service automatically updates the @clientId@ attribute (when empty) for all @Message@ and @PresenceMessage@ messages received from that identified client, and any messages then published from the Ably service, will have the @clientId@ attribute populated. It is therefore expected that Ably client libraries do not explicitly set the @clientId@ field on messages published when @clientId@ is implicit in the connection or authentication scheme
```
Comment (SimonWoolf) [inline content/client-lib-development-guide/features.textile]: `service, will` -> `service will`

`messages published when @clientId@` -> `messages published when the @clientId@`

File: content/client-lib-development-guide/features.textile [hunk 8]
```diff
@@ -165,15 +167,15 @@ h3(#rest-channel). Channel
...
-*** @(RSL1g4)@ When publishing a message with an explicit @clientId@ that is incompatible with the @clientId@ (either inferred or explicitly configured in the token or @ClientOptions@), the library will reject the message immediately and indicate an error
+**** @(RSL1g1b)@ Ably will not assign a @clientId@ upon receiving the @Message@. A test should assert via the history API that the @clientId@ value is @null@ for the @Message@ when received
+*** @(RSL1g2)@ When publishing a @Message@ with the @clientId@ attribute value set to the identified client's @clientId@, Ably will accept the message and publish it. A test should assert that the @clientId@ value is populated for the @Message@ when received
+*** @(RSL1g3)@ When publishing a @Message@ with a different @clientId@ attribute value to the identified client's @clientId@, the client library should reject the message, and indicate an error. The connection and channel remains available for further operations
```
Comment (SimonWoolf) [inline content/client-lib-development-guide/features.textile]: `remains` -> `remain`

**Comment by SimonWoolf** [comment]
:+1:

**Comment by SimonWoolf** [comment]
oops, sorry bout that!

File: content/client-lib-development-guide/features.textile (hunk 1 above)
Comment (mattheworiordan) [inline content/client-lib-development-guide/features.textile]: Sure, that is not what I changed, but I will correct that.

File: content/client-lib-development-guide/features.textile (hunk 2 above)
Comment (mattheworiordan) [inline content/client-lib-development-guide/features.textile]: As above, I just added a basic auth link, but will find all instances now and address in one.

File: content/client-lib-development-guide/features.textile (hunk 6 above)
Comment (mattheworiordan) [inline content/client-lib-development-guide/features.textile]: Sure, thanks, I didn't review the existing copy but simply standardised all references of Token Auth...

File: content/client-lib-development-guide/features.textile (hunk 7 above)
Comment (mattheworiordan) [inline content/client-lib-development-guide/features.textile]: Thanks, will fix now

File: content/client-lib-development-guide/features.textile (hunk 8 above)
Comment (mattheworiordan) [inline content/client-lib-development-guide/features.textile]: Ah, good spot

**Pull request merged**